book/src/template_expansion.md

```diff
@@ -0,0 +1,484 @@
+# Template Expansion
+
+This chapter will explain how the different parts of the templates are
+translated into Rust code.
+
+⚠️ Please note that the generated code might change in the future so the
+following examples might not be up-to-date.
+
+## Basic explanations
+
+Whwn you add `#[derive(Template)]` and `#[template(...)]` on your type, the
+`Template` derive proc-macro will then generate an implementation of the
+`askama::Template` trait which will be a Rust version of the template.
+
+It will also implement the `std::fmt::Display` trait on your type which will
+internally call the `askama::Template` trait.
+
+Let's take a small example:
+
+```rust
+#[derive(Template)]
+#[template(source = "{% set x = 12 %}", ext = "html")]
+struct Mine;
+```
+
+will generate:
+
+```rust
+impl ::askama::Template for YourType {
+    fn render_into(
+        &self,
+        writer: &mut (impl ::std::fmt::Write + ?Sized),
+    ) -> ::askama::Result<()> {
+        let x = 12;
+        ::askama::Result::Ok(())
+    }
+    const EXTENSION: ::std::option::Option<&'static ::std::primitive::str> = Some(
+        "html",
+    );
+    const SIZE_HINT: ::std::primitive::usize = 0;
+    const MIME_TYPE: &'static ::std::primitive::str = "text/html; charset=utf-8";
+}
+
+impl ::std::fmt::Display for YourType {
+    #[inline]
+    fn fmt(&self, f: &mut ::std::fmt::Formatter) -> ::std::fmt::Result {
+        ::askama::Template::render_into(self, f).map_err(|_| ::std::fmt::Error {})
+    }
+}
+```
+
+For simplicity, we will now only keep the content of the `askama::Template::render_into`
+function from now on.
+
+## Text content
+
+If you have "text content" (for example HTML) in your template:
+
+```html
+<h1>{{ title }}</h1>
+```
+
+It will generate it like this:
+
+```rust
+writer
+    .write_fmt(
+        format_args!(
+            "<h1>{0}</h1>",
+            &::askama::MarkupDisplay::new_unsafe(&(self.title), ::askama::Html),
+        ),
+    )?;
+::askama::Result::Ok(())
+```
+
+About `MarkupDisplay`: we need to use this trait in order to prevent generating
+invalid HTML. Let's take an example: if `title` is `"<a>"` and we display it as
+is, in the generated HTML, you won't see `<a>` but instead a new HTML element
+will be created. To prevent this, we need to escape some characters.
+
+In this example, `<a>` will become `&lt;a&gt;`. And this is why there is the
+`safe` builtin filter, in case you want it to be displayed as is.
+
+## Variables
+
+### Variables creation
+
+If you create a variable in your template, it will be created in the generated
+Rust code as well. For example:
+
+```jinja
+{% set x = 12 %}
+{% let y = x + 1 %}
+```
+
+will generate:
+
+```rust
+let x = 12;
+let y = x + 1;
+::askama::Result::Ok(())
+```
+
+### Variables usage
+
+By default, variables will be fields from the type on which the `askama::Template`
+trait is implemented:
+
+```jinja
+{{ y }}
+```
+
+will generate:
```

Comment:
This reads a little awkwardly, you have "implemented: [[ example ]] will generate: [[ generated code ]]". I'd probably expand the "will generate" to something more like "This template will expand as follows:".